platform-grpc-service-framework/src/main/java/org/hypertrace/core/serviceframework/grpc/GrpcPlatformServiceContainer.java

```diff
@@ -38,53 +51,126 @@ public GrpcPlatformServiceContainer(ConfigClient configClient) {
+            .map(
+                entry ->
+                    this.constructServer(
+                        entry.getKey(),
+                        entry.getValue(),
+                        inProcessServerBuilder,
+                        serviceContainerEnvironment))
+            .collect(Collectors.toUnmodifiableList());
+
+    this.servers =
+        Stream.concat(
+                Stream.of(
+                    new ConstructedServer(
+                        this.getInProcessServerName(), inProcessServerBuilder.build())),
+                providedServers.stream())
+            .collect(Collectors.toUnmodifiableList());
+
+    this.healthClient =
+        HealthGrpc.newBlockingStub(this.grpcChannelRegistry.forName(this.getInProcessServerName()));
+  }
+
+  private ConstructedServer constructServer(
+      GrpcPlatformServerDefinition serverDefinition,
+      ServerBuilder<?> networkedBuilder,
+      ServerBuilder<?> inProcessServerBuilder,
+      GrpcServiceContainerEnvironment containerEnvironment) {
+    log.info(
+        "Building server [{}] on port [{}]",
+        serverDefinition.getName(),
+        serverDefinition.getPort());
+    serverDefinition.getServiceFactories().stream()
+        .map(factory -> factory.buildServices(containerEnvironment))
         .flatMap(Collection::stream)
         .map(GrpcPlatformService::getGrpcService)
         .map(InterceptorUtil::wrapInterceptors)
         .forEach(
             service -> {
-              networkedServerBuilder.addService(service);
+              log.info(
+                  "Adding service [{}] to server [{}]",
+                  service.getServiceDescriptor().getName(),
+                  serverDefinition.getName());
+              networkedBuilder.addService(service);
+              log.info(
+                  "Adding service [{}] to in-process server [{}]",
+                  service.getServiceDescriptor().getName(),
+                  getInProcessServerName());
               inProcessServerBuilder.addService(service);
             });
-    inProcessServerBuilder.addService(this.healthStatusManager.getHealthService());
-    this.networkedServer = networkedServerBuilder.build();
-    this.inProcessServer = inProcessServerBuilder.build();
-    this.healthClient =
-        HealthGrpc.newBlockingStub(this.grpcChannelRegistry.forName(this.getInProcessServerName()));
+
+    return new ConstructedServer(serverDefinition.getName(), networkedBuilder.build());
   }
 
   @Override
   protected void doStart() {
-    log.info("Starting: {}", getServiceName());
+    log.info("Starting all services: [{}]", getServiceName());
+    this.startManagedPeriodicTasks();
+    this.servers.forEach(this::startServer);
+    this.servers.forEach(this::awaitServerTermination);
+  }
+
+  private void startManagedPeriodicTasks() {
+    this.periodicTaskExecutor = this.buildTaskExecutor(this.taskDefinitions.size());
+    this.taskDefinitions.forEach(this::startManagedPeriodicTask);
```

Comment:
nit: Would be ideal to clear the taskDefinitions once scheduled.

Reply:
whoops. Initial idea was using the Linked List as a queue to clear it as I read it, but it felt overall less readable so switched back to using it as a list but forgot to clear, fixing.